kaldb/src/main/java/com/slack/kaldb/preprocessor/KaldbSerdes.java

```diff
@@ -0,0 +1,105 @@
+package com.slack.kaldb.preprocessor;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.slack.service.murron.Murron;
+import com.slack.service.murron.trace.Trace;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.serialization.Serializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Factory for creating Kaldb specific serializers / deserializers. Based off of
+ * org.apache.kafka.common.serialization.Serdes
+ */
+public class KaldbSerdes {
+  private static final Logger LOG = LoggerFactory.getLogger(KaldbSerdes.class);
+
+  public static Serde<Murron.MurronMessage> MurronMurronMessage() {
```

Comment:
As discussed this implements the `Serdes` interface and matches the implementations of `org.apache.kafka.common.serialization.Serdes`. It is the caller's responsibility to ensure that ensure that any instrumentation that is required is reported accordingly.

With the streams implementation it instruments both the producer and consumer side of things with a significant amount of metrics, which I've listed below.

<details>
  <summary>Kafka Producer</summary>

```
kafka_producer_connection_close_rate The number of connections closed per second
kafka_producer_iotime_total The total time the I/O thread spent doing I/O
kafka_producer_successful_reauthentication_total The total number of successful re-authentication of connections
kafka_producer_request_latency_max The maximum request latency in ms
kafka_producer_buffer_total_bytes The maximum amount of buffer memory the client can use (whether or not it is currently used).
kafka_producer_response_rate The number of responses received per second
kafka_producer_reauthentication_latency_max The max latency observed due to re-authentication
kafka_producer_bufferpool_wait_time_total The total time an appender waits for space allocation.
kafka_producer_node_request_latency_max 
kafka_producer_io_wait_ratio The fraction of time the I/O thread spent waiting
kafka_producer_failed_reauthentication_total The total number of failed re-authentication of connections
kafka_producer_node_outgoing_byte_total The total number of outgoing bytes
kafka_producer_select_total The total number of times the I/O layer checked for new I/O to perform
kafka_producer_node_request_latency_avg 
kafka_producer_io_wait_time_ns_avg The average length of time the I/O thread spent waiting for a socket ready for reads or writes in nanoseconds.
kafka_producer_record_retry_rate The average per-second number of retried record sends
kafka_producer_buffer_available_bytes The total amount of buffer memory that is not being used (either unallocated or in the free list).
kafka_producer_buffer_exhausted_total The total number of record sends that are dropped due to buffer exhaustion
kafka_producer_records_per_request_avg The average number of records per request.
kafka_producer_network_io_rate The number of network operations (reads or writes) on all connections per second
kafka_producer_record_queue_time_max The maximum time in ms record batches spent in the send buffer.
kafka_producer_record_send_total The total number of records sent.
kafka_producer_produce_throttle_time_max The maximum time in ms a request was throttled by a broker
kafka_producer_node_incoming_byte_total The total number of incoming bytes
kafka_producer_buffer_exhausted_rate The average per-second number of record sends that are dropped due to buffer exhaustion
kafka_producer_bufferpool_wait_ratio The fraction of time an appender waits for space allocation.
kafka_producer_successful_reauthentication_rate The number of successful re-authentication of connections per second
kafka_producer_reauthentication_latency_avg The average latency observed due to re-authentication
kafka_producer_connection_count_total The current number of active connections.
kafka_producer_successful_authentication_total The total number of connections with successful authentication
kafka_producer_node_response_rate The number of responses received per second
kafka_producer_batch_size_avg The average number of bytes sent per partition per-request.
kafka_producer_batch_size_max The max number of bytes sent per partition per-request.
kafka_producer_record_size_avg The average record size
kafka_producer_connection_creation_rate The number of new connections established per second
kafka_producer_io_waittime_total The total time the I/O thread spent waiting
kafka_producer_node_request_total The total number of requests sent
kafka_producer_batch_split_rate The average number of batch splits per second
kafka_producer_failed_reauthentication_rate The number of failed re-authentication of connections per second
kafka_producer_failed_authentication_total The total number of connections with failed authentication
kafka_producer_select_rate The number of times the I/O layer checked for new I/O to perform per second
kafka_producer_node_response_total The total number of responses received
kafka_producer_request_total The total number of requests sent
kafka_producer_record_retry_total The total number of retried record sends
kafka_producer_waiting_threads The number of user threads blocked waiting for buffer memory to enqueue their records
kafka_producer_record_send_rate The average number of records sent per second.
kafka_producer_request_rate The number of requests sent per second
kafka_producer_node_outgoing_byte_rate The number of outgoing bytes per second
kafka_producer_failed_authentication_rate The number of connections with failed authentication per second
kafka_producer_incoming_byte_total The total number of bytes read off all sockets
kafka_producer_io_time_ns_avg The average length of time for I/O per select call in nanoseconds.
kafka_producer_node_incoming_byte_rate The number of incoming bytes per second
kafka_producer_successful_authentication_rate The number of connections with successful authentication per second
kafka_producer_produce_throttle_time_avg The average time in ms a request was throttled by a broker
kafka_producer_outgoing_byte_rate The number of outgoing bytes sent to all servers per second
kafka_producer_request_size_avg The average size of requests sent.
kafka_producer_outgoing_byte_total The total number of outgoing bytes sent to all servers
kafka_producer_request_latency_avg The average request latency in ms
kafka_producer_request_size_max The maximum size of any request sent.
kafka_producer_record_size_max The maximum record size
kafka_producer_record_queue_time_avg The average time in ms record batches spent in the send buffer.
kafka_producer_successful_authentication_no_reauth_total The total number of connections with successful authentication where the client does not support 
kafka_producer_io_ratio The fraction of time the I/O thread spent doing I/O
kafka_producer_network_io_total The total number of network operations (reads or writes) on all connections
kafka_producer_connection_creation_total The total number of new connections established
kafka_producer_node_request_size_max The maximum size of any request sent.
kafka_producer_record_error_total The total number of record sends that resulted in errors
kafka_producer_metadata_age The age in seconds of the current producer metadata being used.
kafka_producer_node_request_size_avg The average size of requests sent.
kafka_producer_response_total The total number of responses received
kafka_producer_requests_in_flight The current number of in-flight requests awaiting a response.
kafka_producer_connection_close_total The total number of connections closed
kafka_producer_compression_rate_avg The average compression rate of record batches.
kafka_producer_batch_split_total The total number of batch splits
kafka_producer_record_error_rate The average per-second number of record sends that resulted in errors
kafka_producer_node_request_rate The number of requests sent per second
kafka_producer_incoming_byte_rate The number of bytes read off all sockets per second
```
</details>

<details>
  <summary>Kafka Consumer</summary>

```
kafka_consumer_coordinator_commit_latency_avg The average time taken for a commit request
kafka_consumer_io_ratio The fraction of time the I/O thread spent doing I/O
kafka_consumer_outgoing_byte_total The total number of outgoing bytes sent to all servers
kafka_consumer_coordinator_commit_total The total number of commit calls
kafka_consumer_coordinator_last_rebalance_seconds_ago The number of seconds since the last successful rebalance event
kafka_consumer_coordinator_join_time_max The max time taken for a group rejoin
kafka_consumer_successful_authentication_rate The number of connections with successful authentication per second
kafka_consumer_failed_reauthentication_rate The number of failed re-authentication of connections per second
kafka_consumer_poll_idle_ratio_avg The average fraction of time the consumer's poll() is idle as opposed to waiting for the user code to process records.
kafka_consumer_coordinator_heartbeat_response_time_max The max time taken to receive a response to a heartbeat request
kafka_consumer_fetch_manager_bytes_consumed_rate The average number of bytes consumed per second
kafka_consumer_fetch_manager_fetch_total The total number of fetch requests.
kafka_consumer_fetch_manager_records_per_request_avg The average number of records in each request
kafka_consumer_io_wait_time_ns_avg The average length of time the I/O thread spent waiting for a socket ready for reads or writes in nanoseconds.
kafka_consumer_successful_authentication_total The total number of connections with successful authentication
kafka_consumer_coordinator_rebalance_latency_total The total number of milliseconds this consumer has spent in successful rebalances since creation
kafka_consumer_time_between_poll_max The max delay between invocations of poll().
kafka_consumer_fetch_manager_fetch_rate The number of fetch requests per second.
kafka_consumer_coordinator_heartbeat_rate The number of heartbeats per second
kafka_consumer_request_size_max The maximum size of any request sent.
kafka_consumer_coordinator_join_total The total number of group joins
kafka_consumer_failed_authentication_rate The number of connections with failed authentication per second
kafka_consumer_coordinator_commit_rate The number of commit calls per second
kafka_consumer_coordinator_rebalance_total The total number of successful rebalance events, each event is composed of several failed re-trials until it succeeded
kafka_consumer_connection_count_total The current number of active connections.
kafka_consumer_request_total The total number of requests sent
kafka_consumer_network_io_total The total number of network operations (reads or writes) on all connections
kafka_consumer_coordinator_join_rate The number of group joins per second
kafka_consumer_outgoing_byte_rate The number of outgoing bytes sent to all servers per second
kafka_consumer_coordinator_assigned_partitions The number of partitions currently assigned to this consumer
kafka_consumer_fetch_manager_records_lead_min The minimum lead in terms of number of records for any partition in this window
kafka_consumer_successful_authentication_no_reauth_total The total number of connections with successful authentication where the client does not support 
kafka_consumer_coordinator_partition_revoked_latency_avg The average time taken for a partition-revoked rebalance listener callback
kafka_consumer_time_between_poll_avg The average delay between invocations of poll().
kafka_consumer_incoming_byte_total The total number of bytes read off all sockets
kafka_consumer_coordinator_failed_rebalance_rate_per_hour The number of failed rebalance events per hour
kafka_consumer_failed_authentication_total The total number of connections with failed authentication
kafka_consumer_reauthentication_latency_max The max latency observed due to re-authentication
kafka_consumer_select_rate The number of times the I/O layer checked for new I/O to perform per second
kafka_consumer_coordinator_sync_total The total number of group syncs
kafka_consumer_connection_close_total The total number of connections closed
kafka_consumer_coordinator_heartbeat_total The total number of heartbeats
kafka_consumer_fetch_manager_bytes_consumed_total The total number of bytes consumed
kafka_consumer_coordinator_failed_rebalance_total The total number of failed rebalance events
kafka_consumer_fetch_manager_fetch_size_avg The average number of bytes fetched per request
kafka_consumer_fetch_manager_records_consumed_rate The average number of records consumed per second
kafka_consumer_coordinator_last_heartbeat_seconds_ago The number of seconds since the last coordinator heartbeat was sent
kafka_consumer_fetch_manager_fetch_size_max The maximum number of bytes fetched per request
kafka_consumer_fetch_manager_fetch_throttle_time_avg The average throttle time in ms
kafka_consumer_iotime_total The total time the I/O thread spent doing I/O
kafka_consumer_successful_reauthentication_rate The number of successful re-authentication of connections per second
kafka_consumer_response_rate The number of responses received per second
kafka_consumer_incoming_byte_rate The number of bytes read off all sockets per second
kafka_consumer_coordinator_rebalance_rate_per_hour The number of successful rebalance events per hour, each event is composed of several failed re-trials until it 
kafka_consumer_fetch_manager_fetch_latency_max The max time taken for any fetch request.
kafka_consumer_coordinator_rebalance_latency_max The max time taken for a group to complete a successful rebalance, which may be composed of several failed re-trials 
kafka_consumer_coordinator_sync_rate The number of group syncs per second
kafka_consumer_request_rate The number of requests sent per second
kafka_consumer_connection_close_rate The number of connections closed per second
kafka_consumer_io_waittime_total The total time the I/O thread spent waiting
kafka_consumer_fetch_manager_fetch_latency_avg The average time taken for a fetch request.
kafka_consumer_fetch_manager_records_lag_max The maximum lag in terms of number of records for any partition in this window
kafka_consumer_request_size_avg The average size of requests sent.
kafka_consumer_coordinator_partition_lost_latency_max The max time taken for a partition-lost rebalance listener callback
kafka_consumer_response_total The total number of responses received
kafka_consumer_network_io_rate The number of network operations (reads or writes) on all connections per second
kafka_consumer_io_wait_ratio The fraction of time the I/O thread spent waiting
kafka_consumer_coordinator_partition_revoked_latency_max The max time taken for a partition-revoked rebalance listener callback
kafka_consumer_select_total The total number of times the I/O layer checked for new I/O to perform
kafka_consumer_coordinator_partition_assigned_latency_avg The average time taken for a partition-assigned rebalance listener callback
kafka_consumer_coordinator_rebalance_latency_avg The average time taken for a group to complete a successful rebalance, which may be composed of several failed re-trials 
kafka_consumer_coordinator_partition_lost_latency_avg The average time taken for a partition-lost rebalance listener callback
kafka_consumer_successful_reauthentication_total The total number of successful re-authentication of connections
kafka_consumer_failed_reauthentication_total The total number of failed re-authentication of connections
kafka_consumer_coordinator_partition_assigned_latency_max The max time taken for a partition-assigned rebalance listener callback
kafka_consumer_fetch_manager_fetch_throttle_time_max The maximum throttle time in ms
kafka_consumer_last_poll_seconds_ago The number of seconds since the last poll() invocation.
kafka_consumer_reauthentication_latency_avg The average latency observed due to re-authentication
kafka_consumer_coordinator_sync_time_max The max time taken for a group sync
kafka_consumer_coordinator_commit_latency_max The max time taken for a commit request
kafka_consumer_coordinator_sync_time_avg The average time taken for a group sync
kafka_consumer_connection_creation_rate The number of new connections established per second
kafka_consumer_connection_creation_total The total number of new connections established
kafka_consumer_fetch_manager_records_consumed_total The total number of records consumed
kafka_consumer_coordinator_join_time_avg The average time taken for a group rejoin
kafka_consumer_io_time_ns_avg The average length of time for I/O per select call in nanoseconds.
```
</details>

<details>
  <summary>Kafka Admin</summary>

```
kafka_stream_thread_poll_total The total number of calls to poll
kafka_stream_thread_punctuate_total The total number of calls to punctuate
kafka_stream_thread_commit_total The total number of calls to commit
kafka_stream_thread_poll_latency_max The maximum poll latency
kafka_stream_thread_punctuate_latency_avg The average punctuate latency
kafka_stream_thread_task_closed_rate The average per-second number of closed tasks
kafka_stream_thread_task_closed_total The total number of closed tasks
kafka_stream_thread_poll_rate The average per-second number of calls to poll
kafka_stream_thread_commit_latency_max The maximum commit latency
kafka_stream_thread_process_total The total number of calls to process
kafka_stream_thread_task_created_rate The average per-second number of newly created tasks
kafka_stream_thread_commit_rate The average per-second number of calls to commit
kafka_stream_thread_punctuate_rate The average per-second number of calls to punctuate
kafka_stream_thread_commit_latency_avg The average commit latency
kafka_stream_thread_task_created_total The total number of newly created tasks
kafka_stream_thread_process_rate The average per-second number of calls to process
kafka_stream_thread_process_latency_avg The average process latency
kafka_stream_thread_poll_latency_avg The average poll latency
kafka_stream_thread_process_latency_max The maximum process latency
kafka_stream_thread_punctuate_latency_max The maximum punctuate latency
```
</details>